File: graphs/cpp/num_islands.cpp
// Each island is a connected component, what we have to find is no. of disconnected components in the graph
// Approach 1 : DFS Traversal

int numIslands(vector<vector<char>>& grid) {
    int islands = 0 ;

    for(int i = 0 ; i < grid.size() ; i++){
        for(int j = 0 ; j < grid[0].size() ; j++){
            if (grid[i][j] == '1'){
                dfs(grid,i,j);
                islands++;
            }
        }
    }
    return islands ;
}


void dfs(vector<vector<char>>&grid, int i, int j){
    if (i < 0 || i >= grid.size() || j < 0 || j >= grid[0].size())
        return ;

    if (grid[i][j]== '2' || grid[i][j] == '0')
        return ;

    grid[i][j] = '2';

    dfs(grid,i+1, j);
    dfs(grid,i, j-1);
    dfs(grid,i-1, j);
    dfs(grid,i, j+1);
}


// Approach2: BFS Traversal: The idea is to find the number of disconnected components

int numIslands(vector<vector<char>>& grid) {
    int n = grid.size() , m = grid[0].size() ;
    int islands = 0 ;

    vector<vector<int>>dir = {{1,0},{0,-1},{-1,0},{0,1}};

    queue<pair<int,int>>q ;
    for(int i = 0 ; i < n ; i++){
        for(int j = 0 ; j < m ; j++){
            if (grid[i][j] == '1'){
                islands++;
                grid[i][j] = '0';
                q.push({i,j});

                while(!q.empty()){
                    auto u = q.front() ;
                    q.pop();

                    for(int k = 0 ; k < 4 ; k++){
                        int x = u.first + dir[k][0];
                        int y = u.second + dir[k][1];

                        if (x >= 0 && x < n && y >=0 && y < m && grid[x][y] == '1'){
                            grid[x][y] = '0';
                            q.push({x,y});
                        }
                    }
                }
            }
        }
    }
    return islands;
}

// TC = O(m*n)
// SC = O(m*n)

File: graphs/go/num_islands.go
// Approach 1 ;
// DFS : use dfs to find the connected components in the graph
// TC = O(n*m)
// SC = O(n*m) as in worst case there might be just a single component

func numIslands(grid [][]byte) int {
	islands := 0
	n := len(grid)
	m := len(grid[0])
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if grid[i][j] == '1' {
				dfs(i, j, grid)
				islands++
			}
		}
	}
	return islands
}

func dfs(i, j int, grid [][]byte) {
	if i < 0 || i >= len(grid) || j < 0 || j >= len(grid[0]) {
		return
	}

	if grid[i][j] == '0' || grid[i][j] == '2' {
		return
	}

	grid[i][j] = '2'

	dfs(i+1, j, grid)
	dfs(i, j-1, grid)
	dfs(i-1, j, grid)
	dfs(i, j+1, grid)
}

// Approach 2 : BFS , use it to find the number of connected components in the graph

type pos struct {
	x int
	y int
}

func numIslands(grid [][]byte) int {
	islands := 0
	n := len(grid)
	m := len(grid[0])

	dir := [][]int{{1, 0}, {0, -1}, {-1, 0}, {0, 1}}

	queue := []pos{}

	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if grid[i][j] == '1' {
				grid[i][j] = '0'
				islands++
				p := pos{
					x: i,
					y: j,
				}
				queue = append(queue, p)

				for len(queue) != 0 {
					item := queue[0]  // remove first
					queue = queue[1:] // reassign to next

					for k := 0; k < 4; k++ {
						x, y := item.x+dir[k][0], item.y+dir[k][1]
						if x >= 0 && x < n && y >= 0 && y < m && grid[x][y] == '1' {
							queue = append(queue, pos{x: x, y: y})
							grid[x][y] = 1
						}
					}
				}
			}
		}
	}
	return islands
}